### PR TITLE
Add STM32 Nucleo L476RG Support

### PR DESCRIPTION
**Pull request by pcting**
Adding support for STM32 L476RG
datasheet: https://www.st.com/resource/en/datasheet/stm32l476je.pdf
board layout: https://www.st.com/resource/en/user_manual/um1724-stm32-nucleo64-boards-mb1136-stmicroelectronics.pdf

Running `blinky3.go` Notes:
* Serial / UART2 works. ~~Serial out prints out garbage~~
* LED blinking works.
* Button works after jumping PC13 PIN to PA0 / A0 PIN.

```bash
$ tinygo flash  -target=nucleo-l476rg src/examples/blinky3/blinky3.go
xPack Open On-Chip Debugger 0.12.0+dev-01557-gdd1758272-dirty (2024-04-02-07:26)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
WARNING: interface/stlink-v2-1.cfg is deprecated, please switch to interface/stlink.cfg
Info : auto-selecting first available session transport "hla_swd". To override use 'transport select <transport>'.
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
Info : clock speed 500 kHz
Info : STLINK V2J33M25 (API v2) VID:PID 0483:374B
Info : Target voltage: 3.258383
Info : [stm32l4x.cpu] Cortex-M4 r0p1 processor detected
Info : [stm32l4x.cpu] target has 6 breakpoints, 4 watchpoints
Info : [stm32l4x.cpu] Examination succeed
Info : starting gdb server for stm32l4x.cpu on 3333
Info : Listening on port 3333 for gdb connections
Info : Unable to match requested speed 500 kHz, using 480 kHz
Info : Unable to match requested speed 500 kHz, using 480 kHz
[stm32l4x.cpu] halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x080031a4 msp: 0x20001000
** Programming Started **
Info : device idcode = 0x10076415 (STM32L47/L48xx - Rev 4 : 0x1007)
Info : RDP level 0 (0xAA)
Info : flash size = 1024 KiB
Info : flash mode : dual-bank
Info : Padding image section 0 at 0x0800c264 with 4 bytes (bank write end alignment)
Warn : Adding extra erase range, 0x0800c268 .. 0x0800c7ff
** Programming Finished **
** Resetting Target **
Info : Unable to match requested speed 500 kHz, using 480 kHz
Info : Unable to match requested speed 500 kHz, using 480 kHz
shutdown command invoked
```

~~it looks like the code flashes fine as the LED blinks every second as expected, but monitoring the serial port outputs garbage. i'm trying to figure out why.~~

building something similar using platformio:
```bash
 *  Executing task: platformio run --target upload 

Processing nucleo_l476rg (platform: ststm32; board: nucleo_l476rg; framework: arduino)
-------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/ststm32/nucleo_l476rg.html
PLATFORM: ST STM32 (17.3.0) > ST Nucleo L476RG
HARDWARE: STM32L476RGT6 80MHz, 96KB RAM, 1MB Flash
DEBUG: Current (stlink) On-board (stlink) External (blackmagic, cmsis-dap, jlink)
PACKAGES: 
 - framework-arduinoststm32 @ 4.20701.0 (2.7.1) 
 - framework-cmsis @ 2.50900.0 (5.9.0) 
 - tool-dfuutil @ 1.11.0 
 - tool-dfuutil-arduino @ 1.11.0 
 - tool-openocd @ 3.1200.0 (12.0) 
 - tool-stm32duino @ 1.0.1 
 - toolchain-gccarmnoneeabi @ 1.120301.0 (12.3.1)
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 14 compatible libraries
Scanning dependencies...
No dependencies
Building in release mode
Checking size .pio/build/nucleo_l476rg/firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [          ]   1.5% (used 1488 bytes from 98304 bytes)
Flash: [          ]   2.2% (used 23240 bytes from 1048576 bytes)
Configuring upload protocol...
AVAILABLE: blackmagic, cmsis-dap, jlink, mbed, stlink
CURRENT: upload_protocol = stlink
Uploading .pio/build/nucleo_l476rg/firmware.elf
xPack Open On-Chip Debugger 0.12.0-01004-g9ea7f3d64-dirty (2023-01-30-15:03)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
debug_level: 1

srst_only separate srst_nogate srst_open_drain connect_deassert_srst

Warn : target stm32l4x.cpu examination failed
[stm32l4x.cpu] halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x08000ff8 msp: 0x20000800
** Programming Started **
Warn : Adding extra erase range, 0x08005c78 .. 0x08005fff
** Programming Finished **
** Verify Started **
** Verified OK **
** Resetting Target **
shutdown command invoked
=============================== [SUCCESS] Took 5.81 seconds ===============================
 *  Terminal will be reused by tasks, press any key to close it. 
```

### CODE REVIEW

File: src/machine/machine_stm32l4x6.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+const APB2_TIM_FREQ = 80e6 // 80MHz
+
+//---------- I2C related code
+
+// Gets the value for TIMINGR register
+func (i2c *I2C) getFreqRange() uint32 {
+	// This is a 'magic' value calculated by STM32CubeMX
+	// for 80MHz PCLK1.
+	// TODO: Do calculations based on PCLK1
+	return 0x10909CEC
```
Comment (pcting) [inline src/machine/machine_stm32l4x6.go:24]: i copied this constant from one of the other stm32 files; hopefully this is right

File: src/machine/board_nucleol476rg.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+)
+
+var (
+	// USART2 is the hardware serial port connected to the onboard ST-LINK
+	// debugger to be exposed as virtual COM port over USB on Nucleo boards.
+	UART1  = &_UART1
+	_UART1 = UART{
+		Buffer:            NewRingBuffer(),
+		Bus:               stm32.USART2,
+		TxAltFuncSelector: 7,
```
Comment (pcting) [inline src/machine/board_nucleol476rg.go]: I could use some insight as to what these numbers represent

File: src/runtime/runtime_stm32l4x6.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	| HCLK        | 80mhz    |
+	| APB1(PCLK1) | 80mhz    |
+	| APB2(PCLK2) | 80mhz    |
+	+-------------+-----------+
+*/
+const (
+	HSE_STARTUP_TIMEOUT = 0x0500
+	PLL_M               = 1
+	PLL_N               = 60
+	PLL_P               = RCC_PLLP_DIV2
```
Comment (pcting) [inline src/runtime/runtime_stm32l4x6.go]: not sure if any of these `PLL_*` constants are correct as i believe these are all incorrect, which is the source of why the serial / UART messages data is garbled.

File: src/machine/board_nucleol476rg.go (hunk 2 above)
Comment (pcting) [reply src/machine/board_nucleol476rg.go]: i think i understand these numbers now as they refer to the "Table 17. Alternate function AF0 to AF7(1) (continued)" section of the datasheet

File: src/runtime/runtime_stm32l4x6.go (hunk 3 above)
Comment (pcting) [reply src/runtime/runtime_stm32l4x6.go]: fixed the issue by copying the values in L4x2. Good follow up reading on the topic: https://community.st.com/t5/stm32-mcus/part-1-introduction-to-the-stm32-microcontroller-clock-system/ta-p/605369

**Comment by pcting** [comment]
occasionally i see this error:

```bash
$ tinygo flash  -target=nucleo-l476rg src/examples/blinky3/blinky3.go
xPack Open On-Chip Debugger 0.12.0+dev-01557-gdd1758272-dirty (2024-04-02-07:26)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
WARNING: interface/stlink-v2-1.cfg is deprecated, please switch to interface/stlink.cfg
Info : auto-selecting first available session transport "hla_swd". To override use 'transport select <transport>'.
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
Info : clock speed 500 kHz
Info : STLINK V2J33M25 (API v2) VID:PID 0483:374B
Info : Target voltage: 3.255227
Info : [stm32l4x.cpu] Cortex-M4 r0p1 processor detected
Info : [stm32l4x.cpu] target has 6 breakpoints, 4 watchpoints
Info : [stm32l4x.cpu] Examination succeed
Info : starting gdb server for stm32l4x.cpu on 3333
Info : Listening on port 3333 for gdb connections
Info : Unable to match requested speed 500 kHz, using 480 kHz
Info : Unable to match requested speed 500 kHz, using 480 kHz
Error: timed out while waiting for target halted
** Unable to reset target **
shutdown command invoked

error: failed to flash /tmp/tinygo3652766739/main.hex: exit status 1
```

## the solutions
### Option 1: Flash Twice Holding Reset
1. while holding down the reset button down, flash the device
2. release the reset button and flash again.

### Option 2: Flash With platform.io
1. create a "hello world" app in platform.io and flash it.

**Comment by deadprogram** [comment]
Hello @pcting thank you for adding this new board/processor.

Did you actually mean to add the additional example? I'm pretty sure that is actually not needed. It would be better to just add the new board to the smoketests located here: https://github.com/tinygo-org/tinygo/blob/dev/GNUmakefile#L696-L731

**Comment by pcting** [comment]
> Hello @pcting thank you for adding this new board/processor.
> 
> Did you actually mean to add the additional example? I'm pretty sure that is actually not needed. It would be better to just add the new board to the smoketests located here: https://github.com/tinygo-org/tinygo/blob/dev/GNUmakefile#L696-L731

ah, i wasn't aware of the `smoketest`  target. thanks! i'll remove the new blinky3.go example

**Comment by pcting** [comment]
@deadprogram i've cleaned up the unneeded example. let me know if anything else needs changing. thanks

File: .gitignore [hunk 4]
```diff
@@ -32,3 +32,7 @@ test.hex
 test.nro
 test.wasm
 wasm.wasm
+
+# Ignore vscode
+.vscode
+.tool-versions
```
Comment (sago35) [inline .gitignore]: This change was discussed in the following Issue. As it is an unrelated change to this PR, please revert it.

 #3835

**Review by sago35 — CHANGES_REQUESTED** [review]
Thank you for creating the PR. I think it's a very good code with just a few minor points to address. By reverting the changes in .gitignore, I believe this PR can be merged.

File: .gitignore (hunk 4 above)
Comment (pcting) [reply .gitignore]: done

**Comment by pcting** [comment]
i think i need to revisit the arduino pin mapping before we merge

**Comment by pcting** [comment]
i fixed the pin mapping; it should be good now

File: targets/nucleo-l476rg.json [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+    "build-tags": ["nucleol476rg", "stm32l476", "stm32l4x6", "stm32l4", "stm32"],
+    "serial": "uart",
+    "linkerscript": "targets/stm32l4x6.ld",
+    "extra-files": [
+      "src/device/stm32/stm32l4x6.s"
+    ],
+    "flash-method": "openocd",
+    "openocd-interface": "stlink-v2-1",
+    "openocd-target": "stm32l4x"
+  }
```
Comment (sago35) [inline targets/nucleo-l476rg.json]: Please add LF at the end of the file. I think it's better to keep it consistent with other files.

File: targets/stm32l4x6.ld [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+MEMORY
+{
+  FLASH_TEXT (rx) : ORIGIN = 0x08000000, LENGTH = 1024K
+  RAM (xrw)       : ORIGIN = 0x20000000, LENGTH = 96K
+  RAM2 (xrw)      : ORIGIN = 0x10000000, LENGTH = 32K
+}
+
+_stack_size = 4K;
+
+INCLUDE "targets/arm.ld"
```
Comment (sago35) [inline targets/stm32l4x6.ld]: Please add LF at the end of the file. I think it's better to keep it consistent with other files.

**Review by sago35 — CHANGES_REQUESTED** [review]
It's almost LGTM, but please add the line break code. With that, it should be mergeable this time.

File: targets/stm32l4x6.ld (hunk 6 above)
Comment (pcting) [reply targets/stm32l4x6.ld]: done

File: targets/nucleo-l476rg.json (hunk 5 above)
Comment (pcting) [reply targets/nucleo-l476rg.json]: done

**Review by sago35 — APPROVED** [review]
LGTM

**Comment by deadprogram** [comment]
Thanks again @pcting and @sago35 now merging.
